tensorflow/core/kernels/dml_segment_reduction_ops.cc

```diff
@@ -0,0 +1,255 @@
+
+#include <numeric>
+
+#include "tensorflow/core/common_runtime/dml/dml_operator_helper.h"
+#include "tensorflow/core/common_runtime/dml/dml_util.h"
+#include "tensorflow/core/framework/bounds_check.h"
+#include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/kernels/dml_kernel_wrapper.h"
+#include "tensorflow/core/kernels/dml_ops_common.h"
+
+namespace tensorflow {
+
+class SegmentReductionInitializationHelper : public InitializationHelper {
+ public:
+  using Attributes = EmptyAttributes;
+
+  SegmentReductionInitializationHelper(OpKernelContext* ctx,
+                                       std::shared_ptr<const Attributes> attr) {
+    const Tensor& data = ctx->input(0);
+    const Tensor& segment_ids = ctx->input(1);
+    const Tensor& num_segments = ctx->input(2);
+
+    OP_REQUIRES(
+        ctx, TensorShapeUtils::IsScalar(num_segments.shape()),
+        errors::InvalidArgument("num_segments should be a scalar, not shape ",
+                                num_segments.shape().DebugString()));
+    OP_REQUIRES(
+        ctx, TensorShapeUtils::StartsWith(data.shape(), segment_ids.shape()),
+        errors::InvalidArgument("data.shape = ", data.shape().DebugString(),
+                                " does not start with segment_ids.shape = ",
+                                segment_ids.shape().DebugString()));
+
+    const int64 output_rows = internal::SubtleMustCopy(static_cast<int64>(
+        num_segments.dtype() == DT_INT32 ? num_segments.scalar<int32>()()
+                                         : num_segments.scalar<int64>()()));
+    OP_REQUIRES(ctx, output_rows >= 0,
+                errors::InvalidArgument("Input num_segments == ", output_rows,
+                                        " must not be negative."));
+
+    output_shape_.AddDim(output_rows);
+    for (int i = segment_ids.dims(); i < data.dims(); i++) {
+      output_shape_.AddDim(data.dim_size(i));
+    }
+  }
+
+  const TensorShape& GetOutputShape() const { return output_shape_; }
+
+ private:
+  TensorShape output_shape_;
+};
+
+class SegmentReductionShapeHelper : public ShapeHelper {
+ public:
+  std::vector<TensorShape> GetOutputShapes(
+      OpKernelContext* ctx,
+      const InitializationHelper* initialization_helper) const override {
+    auto init_helper = static_cast<const SegmentReductionInitializationHelper*>(
+        initialization_helper);
+
+    return {init_helper->GetOutputShape()};
+  }
+};
+
+template <DML_REDUCE_FUNCTION reduce_function, typename TData>
+constexpr TData GetIdentityValue() {
+  if (reduce_function == DML_REDUCE_FUNCTION_MULTIPLY) {
+    return static_cast<TData>(1);
+  }
+
+  if (reduce_function == DML_REDUCE_FUNCTION_SUM) {
+    return static_cast<TData>(0);
+  }
+
+  if (reduce_function == DML_REDUCE_FUNCTION_MIN) {
+    return std::numeric_limits<TData>::max();
+  }
+
+  if (reduce_function == DML_REDUCE_FUNCTION_MAX) {
+    return std::numeric_limits<TData>::lowest();
+  }
+}
```

Comment:
Might want to assert or abort or something in case somebody accidentally supplies a reduce function that isn't one of these four.

Reply:
It won't compile then because it's a constexpr function and needs to return something at compile-time.

Reply:
It's not as easy to read, but you can use template specialization (with classes not functions):

```
template <DML_REDUCE_FUNCTION reduce_function, typename TData> 
struct IdentityValueFunctor { constexpr TData operator()() = delete; };

template <typename TData>
struct IdentityValueFunctor<DML_REDUCE_FUNCTION_MULTIPLY, TData>
{ constexpr TData operator()() { return static_cast<TData>(1); } };

template <typename TData>
struct IdentityValueFunctor<DML_REDUCE_FUNCTION_SUM, TData>
{ constexpr TData operator()() { return static_cast<TData>(0); } };

template <typename TData>
struct IdentityValueFunctor<DML_REDUCE_FUNCTION_MIN, TData>
{ constexpr TData operator()() { return std::numeric_limits<TData>::max(); } };

template <typename TData>
struct IdentityValueFunctor<DML_REDUCE_FUNCTION_MAX, TData>
{ constexpr TData operator()() { return std::numeric_limits<TData>::lowest(); } };
```